### ISSUE TITLE
[distributed] fallocate/fadvise to provide hint for I/O

### ISSUE THREAD

**Issue by harshavardhana** [minio/minio#2194]
Under `CompleteMultipartUpload` we should use fallocate/fadvise when available to allocate disk space.

**Comment by vadmeste** [comment]
@harshavardhana, I decided to write a separate code which, at the beginning, will use standard Read/Write routines as used in Minio, then we will try to optimize it for Linux at least.

**Comment by vadmeste** [comment]
Basically, the use case will be merging thousand of 5mb files into one 5gb file, I will target this case then merge it to Minio.

**Comment by vadmeste** [comment]
Using sendfile() speeds up the merging of files, no memory copy with this syscall but it should be much beneficial when Minio is under stress  

```
go read/write: 0m25.275s  0m25.261s  0m25.553s  
sendfile: 0m22.675s  0m23.563s  0m22.706s
```

**Comment by vadmeste** [comment]
Upon my current understanding, fallocate doesn't accelerate copying, it only makes us sure that we won't get a disk full error when copying. fadvise could be beneficial if we have other files that we are recurrently reading and we don't like to invalidate them from the cache buffer, so its effects could appear with a high server overload. Also I tried readahead() and  POSIX_FADV_SEQUENTIAL but it didn't really show any speed enhancement for a normal test (not parallel)

sendfile seems to be promising and could be used to read files from disk and send them over network with very low overhead (zero copy)

**Comment by harshavardhana** [comment]
> Upon my current understanding, fallocate doesn't accelerate copying, it only makes us sure that we won't get a disk full error when copying. fadvise could be beneficial if we have other files that we are recurrently reading and we don't like to invalidate them from the cache buffer, so its effects could appear with a high server overload. Also I tried readahead() and POSIX_FADV_SEQUENTIAL but it didn't really show any speed enhancement for a normal test (not parallel)
> 
> sendfile seems to be promising and could be used to read files from disk and send them over network with very low overhead (zero copy)

We should still implement `fallocate` when available, allows for efficient usage of storage. you can go ahead and implement it for FS backend.

**Comment by GarimaKapoor** [comment]
please update the timeline for the fix!

**Comment by vadmeste** [comment]
This only needs to be merged

**Comment by vadmeste** [comment]
Well, for the moment, I added AppendFile() to the storage layer to call fallocate before starting to writing in the object. For the moment, this is tested only with FS backend and with no multi-parted objects.

I had the idea to see filesystem non-contiguous percentage before measuring speed performance to see difference between using fallocate and not using it. I used two parallel mc to upload multiple files of 50 mb (< 64mb) until the disk becomes full. This is the comparison:
- Fresh disk, no uploaded files, **0.0% contiguous** 
- Fresh disk, many 50 mb files uploaded by two parallel mc: **43.0% non-contiguous**
- Fresh disk, many 50 mb files uploaded by two parallel mc with **falloc** activated: **8.1% non-contiguous**

**Comment by krishnasrinivas** [comment]
@vadmeste did you check the time difference between the two configs?

also how did you find out contiguous/non-contiguous %?

**Comment by vadmeste** [comment]
> also how did you find out contiguous/non-contiguous %?

well, I umount my disk then I run "fsck -nf /dev/mydisk" to see non contiguous percentage.

> did you check the time difference between the two configs?

I am doing it right now.. I probably need to disable some memory caching first.. will update soon

**Comment by harshavardhana** [comment]
> Well, for the moment, I added AppendFile() to the storage layer to call fallocate before starting to writing in the object. For the moment, this is tested only with FS backend and with no multi-parted objects.
> 
> I had the idea to see filesystem non-contiguous percentage before measuring speed performance to see difference between using fallocate and not using it. I used two parallel mc to upload multiple files of 50 mb (< 64mb) until the disk becomes full. This is the comparison:
> 
> Fresh disk, no uploaded files, 0.0% contiguous
> Fresh disk, many 50 mb files uploaded by two parallel mc: 43.0% non-contiguous
> Fresh disk, many 50 mb files uploaded by two parallel mc with falloc activated: 8.1% non-contiguous

:+1:  - This is great to know our previous theory is nullified. Can you validate this with XFS as well?

**Comment by vadmeste** [comment]
> - This is great to know our previous theory is nullified. Can you validate this with XFS as well?

Well, XFS is known to optimize fragmentation, so I decided to run 4 parallel mc instead of 2. Here is my findings:
- without fallocate:  **16.88%**
- with fallocate: **0.76%**

**Comment by harshavardhana** [comment]
> Well, XFS is known to optimize fragmentation, so I decided to run 4 parallel mc instead of 2. Here is my findings:
> 
> without fallocate: 16.88%
> with fallocate: 0.76%

Still better XFS starts showing bad signs as the usage gets higher on large disks.  So it is good to know fallocate helps.

**Comment by GarimaKapoor** [comment]
@vadmeste please update the status of this issue?

**Comment by vadmeste** [comment]
Timeline: 21 Oct.

**Comment by krishnasrinivas** [comment]
> This is great to know our previous theory is nullified. Can you validate this with XFS as well?

previously we were doing fallocate in 128K chunks in AppendFile (which will not add value), now @vadmeste is doing fallocate with PrepareFile() StorageAPI allocating the space for the entire file.

**Comment by harshavardhana** [comment]
Fixed by #3032
